### PR TITLE
[TG-8445] Dynamically allocate arguments to java entry point

### PR DESCRIPTION
**Pull request by owen-mc-diffblue**
This was already the case for arguments that had more than
one object type. This PR makes it the case for arguments
that only have one object type too.

The motivation for this is that when nondet-initialising the
argument with automatic local allocation we were creating the
object to be pointed to and only initialising it when the
argument is non-null, and then symex couldn't resolve the
class identifier, which stopped it from doing various
optimisations and producing a simpler formula.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by smowton** [comment]
The ideal test for this should exhibit its expected benefit -- that a classid test against something so allocated can be resolved by symex, and so a subsequent virtual call only explores the really-reachable version of the function.

File: jbmc/src/java_bytecode/java_entry_point.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_entry_point.cpp:354]: :bulb: I wonder if now all of the objects are dynamic if we can simplify the `allocate_objectst` to just be a function

**Review by thk123 — COMMENTED** [review]
As Chris says, ideally we need some kind of test. Also, this would benefit from a run of test-gens tests before merging, I'll can do this tomorrow unless you beat me to it :slightly_smiling_face:

**Comment by thk123** [comment]
TG looks OK - but this needs a rebase and to pass CI here first...

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d3bc7bd2f78e0fea18a8f88929718b13a93a129c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118108484

File: jbmc/src/java_bytecode/java_entry_point.cpp
Comment (owen-mc-diffblue) [reply jbmc/src/java_bytecode/java_entry_point.cpp:354]: I don't know anything about it. If so, that could be a follow-up PR.

**Comment by owen-mc-diffblue** [comment]
@thk123 @smowton I've fixed CI and added a test, so this is ready to review.

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: `irep_idt identifier;` or `irep_idt identifier{};` achieves the same result

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: Again no need for init

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: This makes the function a bit weird-- we're happy with `x.a = b;` and `(*x).a = b;` aka `x->a = b;`? The first one assigns (a component of) `x`, the second one looks through it to assign something else... that suggests the code *calling* this function has the wrong idea about what it's looking for.

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: like what?

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: There are only 3 cases listed in the comment above

**Review by smowton — COMMENTED** [review]
It's a mess, what with the client unit tests clearly looking for something much more specific than is warranted... I'd suggest something testing all of `java_entry_point` should be checking for its functional properties -- that it produces initial assignments capable of satisfying constraints X Y Z but not A B C -- rather than that it produces precise output code (that would be appropriate for code testing `java_object_factory` specifically in `AUTOMATIC_LOCAL` mode). Can we just turn these tests into regression tests?

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (allredj) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp]: Don't some compilers require that? (cf https://github.com/diffblue/cbmc/pull/4876)

Comment (tautschnig) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp]: I am still puzzled about #4876. Maybe clang++-3.7 is completely broken?

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (owen-mc-diffblue) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp]: Yes, it is weird. Maybe I should make a separate function for the `x->a` case.

Comment (owen-mc-diffblue) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp]: It turns out that we no longer need the `x.a` case, so I've fully switched it over to `x->a`.

**Comment by owen-mc-diffblue** [comment]
@smowton This is ready to re-review

File: jbmc/unit/java_bytecode/goto_program_generics/generic_bases_test.cpp
Comment (smowton) [inline jbmc/unit/java_bytecode/goto_program_generics/generic_bases_test.cpp]: Surprising to see the comments changing but not the code?

Comment (smowton) [reply jbmc/unit/java_bytecode/goto_program_generics/generic_bases_test.cpp]: Also surely we always expect both a declaration and an allocation, not either/or?

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: In the has not case are we expecting it is nulled?

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: Resulting statement = `assignment`?

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: I'm not sure either the name or comments match what this does? It appears to expect an assignment a = b, then go looking for either b = ALLOCATE(...) or b = c, where c is again a symbol? Either way I can't see how struct_component_ptr_types are involved?

**Review by smowton — COMMENTED** [review]
Certainly much tidier than before, but I'd appreciate some explanation regarding what we're looking for in the modified unit test

File: jbmc/unit/java_bytecode/goto_program_generics/generic_bases_test.cpp
Comment (owen-mc-diffblue) [reply jbmc/unit/java_bytecode/goto_program_generics/generic_bases_test.cpp]: The comment is giving what code we expect to find. That has changed because `this_tmp_name` comes `require_entry_point_argument_assignment()`, and entry point arguments are now assigned differently.

Comment (owen-mc-diffblue) [reply jbmc/unit/java_bytecode/goto_program_generics/generic_bases_test.cpp]: Yes. I should make the comment clearer about what's going on. Also, I think I'm trying to explain too much in the wrong place.

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (owen-mc-diffblue) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp]: I'll expand that comment to make things clearer. It should say that we're expecting a symbol has been assigned to it, and it's that symbol which we return. In the short term it doesn't matter which symbol we use as they both have the same type, but we return the symbol and some tests continue to use it for other things, and in case 2 they need the second symbol.

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (owen-mc-diffblue) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp]: Yes. I'll fix that. I think I moved this comment from somewhere else, where that phrasing presumably made more sense.

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (owen-mc-diffblue) [reply jbmc/unit/java-testing-utils/require_goto_statements.cpp]: Your summary is correct, except it's `a->x = b`. The fact that the returned symbol (`b` or `c`) has the correct type comes after the call to `get_name_of_symbol_with_struct_component_ptr_type`. So it's more like `get_name_of_symbol_which_we_believe_will_have_struct_component_ptr_type`. The function doesn't check that it has that type in any way.

**Comment by owen-mc-diffblue** [comment]
I've addressed @smowton's comments, which helped me think more about what's going on and make it much clearer.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 11de12c3394a5f7664e8325cb55183ed1336ed37).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118588948

**Comment by allredj** [comment]
This looks like it takes code coverage down a bit. Do you see why?

**Comment by owen-mc-diffblue** [comment]
I had a quick look (without really knowing how to use codecov.io) and didn't find anything that seemed bad. Here's my summary of the biggest changes:

Lost 5 lines of coverage in `exprt boolbvt::bv_get_unbounded_array(const exprt &expr) const` because it is no longer called with values.size() finite and known in advance

In `src/statement-list/statement_list_typecheck.cpp` various typecheck functions are no longer called, presumably because the things they check for are no longer created.
```
typecheck_statement_list_load(op_code, tia_element);
typecheck_statement_list_transfer(op_code, tia_element);
typecheck_statement_list_accu_int_add(op_code);
typecheck_statement_list_accu_int_sub(op_code);
typecheck_statement_list_accu_int_mul(op_code);
typecheck_statement_list_accu_int_div(op_code);
typecheck_statement_list_accu_real_mul(op_code);
typecheck_statement_list_accu_real_div(op_code);
typecheck_statement_list_set_rlo(op_code);
```

`static void get_sub_arrays(const exprt &array_expr, std::vector<exprt> &accu)` seems to now only be called with array_exprts now, not if_exprts or anything expression which contains an array_exprt as an operand.

**Comment by owen-mc-diffblue** [comment]
To some extent changes like this are to be expected: this PR means that we allocate some objects in different ways, producing different code, which then exercises different code paths. I'm not sure we should remove any of the lines which are not covered any more, because they might well be useful in future.

(This is all assuming that the results of  codecov.io are fully reliable.)

**Comment by allredj** [comment]
@owen-jones-diffblue OK, thanks for looking into this. It seems fine then! :+1:

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: I think these case labels are the wrong way around? Case 2 is when there *is* a source symbol

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: How about get_ultimate_source_symbol?

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp:331]: This would be easier to explain if split into three functions: one that finds a unique non-null *expression* feeding a symbol, one that finds a unique *symbol*, and one which does the iteration. Then the first can be explained tersely that the input program may contain many "input = null" assignments and must contain exactly one "input = e"; returns e. The others are then easier to define in terms of that operation.

**Review by smowton — COMMENTED** [review]
Apologies for being a pain in the arse on this one, but I'd really like to come up with a clearer formulation for this function if possible, and I suspect the way to do that is to split it into smaller functions with simpler individual responsibilities.

File: jbmc/unit/java_bytecode/goto_program_generics/mutually_recursive_generics.cpp
Comment (majakusber) [inline jbmc/unit/java_bytecode/goto_program_generics/mutually_recursive_generics.cpp:20]: :pick: Why are these needed? Especially, why is the one here required?

**Review by majakusber — APPROVED** [review]
Looks good to me :+1:

File: jbmc/unit/java_bytecode/goto_program_generics/mutually_recursive_generics.cpp
Comment (owen-mc-diffblue) [reply jbmc/unit/java_bytecode/goto_program_generics/mutually_recursive_generics.cpp:20]: Without it there is failure when simplifying 0+0, because certain indexes have width 0 and the simplify_plus method checks the widths are non-zero. I assume it is needed now because some things became pointers that weren't before.

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: This doesn't enforce that it's a dynamic allocation, so just say it's either copied from another symbol (in which case we return that symbol expression) or from something else (when we return null)

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: "is always" -> "must be"

File: jbmc/unit/java-testing-utils/require_goto_statements.cpp
Comment (smowton) [inline jbmc/unit/java-testing-utils/require_goto_statements.cpp]: Give a short example, like 
```
a = 5 + g(7)
b = a
c = b
// Given input c, returns a
```

**Review by smowton — APPROVED** [review]
Just nitpicks now, thanks for improving this!

**Comment by owen-mc-diffblue** [comment]
TG bump: https://github.com/diffblue/test-gen/pull/3681

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 81b7e7c091c0d7f7560b1cdaf57cb69fa321d5a4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119131808

**Comment by owen-mc-diffblue** [comment]
I will merge this when I've got the TG bump passing

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2c36f47d7d7c173b7e40b30e9bdbc790187a21a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119297350

**Pull request merged**
